# ML-08 — Capstone Modeling Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/VipS-2004/flyrank1/blob/main/work/notebooks/w05_model.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [4]:
!git clone https://github.com/VipS-2004/flyrank1.git

fatal: destination path 'flyrank1' already exists and is not an empty directory.


## 1. Method choice and why

*Which method from the toolkit, and why it fits your lane.*

Method choice: I chose correlation and signal analysis because the Week-4 baseline is a heuristic ranking score rather than a supervised predictive model, and the dataset does not contain a clear ground-truth label for whether a content refresh was successful. Correlation analysis lets me investigate which available content and performance signals are associated with engagement and whether the signals used by the baseline are actually supported by the data. This is more appropriate than training a classifier on the baseline score itself, which would be circular.

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. Split design

*Grouped by client? Time-aware? Say why this split is honest for your question.*

Split design: This analysis does not train a predictive model, so a train/test split is not required. I will use the full dataset for exploratory correlation and signal analysis. I will avoid using the baseline_score, reason_code, and action columns as explanatory signals because they are derived from the Week-4 heuristic itself and would create circular analysis

In [6]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Train + compare vs my baseline

*Same data, same metric, same split as your Week-4 baseline. Show the table.*

In [7]:
import pandas as pd
import numpy as np

# Load the same Week-4 dataset
df = pd.read_csv("/content/flyrank1/data/raw/content_refresh_anonymized.csv")

# Signals that are independent of the Week-4 heuristic
signal_cols = [
    "content_age_days",
    "days_since_last_update",
    "impressions_90d",
    "clicks_90d",
    "sessions_90d",
    "ai_traffic_pct",
    "word_count",
    "ctr",
    "avg_position",
    "engagement_rate",
    "scroll_rate",
    "trend_pct"
]

# Correlation of each signal with engagement rate
correlations = (
    df[signal_cols]
    .corr(numeric_only=True)["engagement_rate"]
    .drop("engagement_rate")
    .sort_values(key=abs, ascending=False)
)

print("Signal correlation with engagement_rate:")
display(correlations.to_frame("correlation"))

# Week-4 baseline summary
df["baseline_score"] = (
    0.4 * (df["content_age_days"] / df["content_age_days"].max())
    + 0.3 * (1 - df["ctr"] / 100)
    + 0.3 * (1 - df["engagement_rate"] / 100)
)

# Week-4 baseline summary
baseline_summary = pd.DataFrame({
    "Method": ["Week-4 Heuristic Baseline"],
    "Metric": ["Mean baseline score"],
    "Value": [df["baseline_score"].mean()]
})

display(baseline_summary)

Signal correlation with engagement_rate:


,correlation
scroll_rate,0.162611
ctr,0.096903
word_count,-0.063045
content_age_days,0.039776
ai_traffic_pct,0.031504
clicks_90d,0.030527
impressions_90d,0.024318
avg_position,-0.017926
days_since_last_update,-0.013661
trend_pct,0.008412


,Method,Metric,Value
0,Week-4 Heuristic Baseline,Mean baseline score,0.772544


In [8]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 4. Errors and interpretation

*Where is the model wrong? What does it lean on? A short error analysis beats a big metric table.*

Errors and interpretation: The analysis does not produce prediction errors because it is a correlation-based signal analysis rather than a trained predictive model. The strongest observed relationship with engagement rate is scroll_rate (correlation ≈ 0.163), followed by ctr (≈ 0.097). These are relatively weak correlations, so neither signal should be treated as a strong predictor on its own. word_count shows a small negative relationship (≈ -0.063). The Week-4 heuristic baseline has a mean score of 0.773, but this score is a ranking heuristic rather than a measured prediction accuracy. Overall, the results suggest that the baseline's signals should be interpreted cautiously and that engagement is influenced by multiple factors.

In [9]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.